### PR TITLE
SpendBySupplier report with selectable date range using LedgerPack

### PR DESCRIPTION
**Pull request by RossStewart**
Split into separate report as per Andrew McV's request. Simplified version of the report but can search between dates. Uses LedgerPack to determine correct period.

### CODE REVIEW

File: src/Domain.LinnApps/Reports/SpendsReportService.cs [hunk 1]
```diff
@@ -113,6 +118,65 @@ public ResultsModel GetSpendBySupplierReport(string vendorManagerId)
...
+                $"Spend by supplier report for Vendor Manager: {vendorManagerName} between ledger period {fromLedgerPeriod} to {toLedgerPeriod}.");
+
+            AddSupplierByDateRangeReportColumns(reportLayout);
+
+            var values = new List<CalculationValueModel>();
+
+            var distinctSupplierSpends = supplierSpends.DistinctBy(x => x.SupplierId).Select(
+                x => new SupplierSpendWithTotals
+                         {
+                             BaseTotal = x.BaseTotal.HasValue ? x.BaseTotal.Value : 0,
```
Comment (lewisrenfrew) [inline src/Domain.LinnApps/Reports/SpendsReportService.cs:156]: .GetValueOrDefault() (Or megrging the conditional) is cleaner for all of these nullable fields 'value or zero' style checks

File: src/Service.Host/client/src/components/reports/SpendBySupplierByDateRange.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+} from '@linn-it/linn-form-components-library';
+import Grid from '@mui/material/Grid';
+import { useSelector, useDispatch } from 'react-redux';
+import queryString from 'query-string';
+import history from '../../history';
+import config from '../../config';
+import spendBySupplierByDateRangeActions from '../../actions/spendBySupplierByDateRangeActions';
+
+function SpendBySupplierByDateRangeReport() {
+    const options = useMemo(() => queryString.parse(window.location.search) || {}, []);
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/reports/SpendBySupplierByDateRange.js:17]: useLocation() is clearer/less code

File: src/Service.Host/client/src/components/reports/SpendBySupplierByDateRange.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    }, [options, dispatch]);
+
+    return (
+        <>
+            <Grid style={{ marginTop: 40 }} container spacing={3} justifyContent="center">
+                <Grid item xs={12}>
+                    <BackButton backClick={() => handleBackClick(history, options)} />
+                </Grid>
+                <Grid item xs={12}>
+                    {!loading && reportData ? (
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/reports/SpendBySupplierByDateRange.js:46]: you could save a few lines by doing  {!loading && reportData && <stuff to render />} as opposed to this conditional ? thing to save you rendering that empty Fragment

File: src/Service.Host/client/src/components/reports/SpendBySupplierByDateRangeOptions.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                                            }
+                                            return 0;
+                                        })
+                                        .map(v => ({
+                                            id: v.vmId,
+                                            displayText: `${v.vmId} ${v.name} (${v.userNumber})`
+                                        }))
+                                ]}
+                                allowNoValue={false}
+                                onChange={(propertyName, newValue) => setVm(newValue)}
```
Comment (lewisrenfrew) [inline src/Service.Host/client/src/components/reports/SpendBySupplierByDateRangeOptions.js:91]: propertyName is unused here so the linter might be complaining?

File: src/Service/Modules/Reports/SpendsReportModule.cs [hunk 5]
```diff
@@ -82,5 +85,50 @@ private async Task GetApp(HttpRequest req, HttpResponse res)
...
+            var options = new SpendBySupplierByDateRangeReportRequestResource
+            {
+                VendorManager = vm ?? string.Empty,
+                FromDate = fromDate,
+                ToDate = toDate,
+            };
+
+            var csv = spendsReportFacadeService.GetSpendBySupplierByDateRangeReportExport(options);
+
+            var contentDisposition = new ContentDisposition
```
Comment (lewisrenfrew) [inline src/Service/Modules/Reports/SpendsReportModule.cs:106]: what is this variable for?

File: tests/Integration/Integration.Tests/SpendsReportModuleTests/WhenGettingSpendBySupplierByDateRangeExport.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+            this.Response = this.Client.Get(
+                "/purchasing/reports/spend-by-supplier-by-date-range/export?vm=&fromDate=&toDate=",
+                with => { with.Accept("text/csv"); }).Result;
+        }
+
+        [Test]
+        public void ShouldCallDomainService()
+        {
+            this.DomainService.Received().GetSpendBySupplierByDateRangeReport(string.Empty, string.Empty, string.Empty);
```
Comment (lewisrenfrew) [inline tests/Integration/Integration.Tests/SpendsReportModuleTests/WhenGettingSpendBySupplierByDateRangeExport.cs:33]: you might as well pass options in the url and make sure the correct things get passed the domain here (as opposed to just empty strings)

**Review by lewisrenfrew — APPROVED** [review]

File: src/Service.Host/client/src/components/reports/SpendBySupplierByDateRange.js (hunk 2 above)
Comment (richardiphillips) [reply src/Service.Host/client/src/components/reports/SpendBySupplierByDateRange.js:17]: +1 from me. would like to eliminate all the window.location.search from all our react code

Comment (RossStewart) [reply src/Service.Host/client/src/components/reports/SpendBySupplierByDateRange.js:17]: Made the changes. Will remove window.location.search the morning I'm back, test it and we can merge this and check it off the Purchasing Project.

Comment (RossStewart) [reply src/Service.Host/client/src/components/reports/SpendBySupplierByDateRange.js]: Scratch that, just fixed it before I go. All working now. Will resolve and close the other ticket. Just needs added to the menu.

**Pull request merged**